src/crypto/x509/root_windows.go

```diff
@@ -226,41 +231,78 @@ func (c *Certificate) systemVerify(opts *VerifyOptions) (chains [][]*Certificate
 }
 
 func loadSystemRoots() (*CertPool, error) {
-	// TODO: restore this functionality on Windows. We tried to do
-	// it in Go 1.8 but had to revert it. See Issue 18609.
-	// Returning (nil, nil) was the old behavior, prior to CL 30578.
-	return nil, nil
+	// // TODO: restore this functionality on Windows. We tried to do
+	// // it in Go 1.8 but had to revert it. See Issue 18609.
+	// // Returning (nil, nil) was the old behavior, prior to CL 30578.
+	// return nil, nil
+
+	// const CRYPT_E_NOT_FOUND = 0x80092004
+
+	// store, err := syscall.CertOpenSystemStore(0, syscall.StringToUTF16Ptr("ROOT"))
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// defer syscall.CertCloseStore(store, 0)
+
+	// roots := NewCertPool()
+	// var cert *syscall.CertContext
+	// for {
+	// 	cert, err = syscall.CertEnumCertificatesInStore(store, cert)
+	// 	if err != nil {
+	// 		if errno, ok := err.(syscall.Errno); ok {
+	// 			if errno == CRYPT_E_NOT_FOUND {
+	// 				break
+	// 			}
+	// 		}
+	// 		return nil, err
+	// 	}
+	// 	if cert == nil {
+	// 		break
+	// 	}
+	// 	// Copy the buf, since ParseCertificate does not create its own copy.
+	// 	buf := (*[1 << 20]byte)(unsafe.Pointer(cert.EncodedCert))[:]
+	// 	buf2 := make([]byte, cert.Length)
+	// 	copy(buf2, buf)
+	// 	if c, err := ParseCertificate(buf2); err == nil {
+	// 		roots.AddCert(c)
+	// 	}
+	// }
+	// return roots, nil
+
+	tempdir, err := ioutil.TempDir("", "go_x509_certs")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tempdir)
 
-	const CRYPT_E_NOT_FOUND = 0x80092004
+	cmdOut, err := exec.Command("certutil", "-syncWithWU", "-f", tempdir).Output()
```

Comment:
Why do you think that calling certutil.exe won't have problem described in https://github.com/golang/go/issues/18609 ?

Alex

Reply:
As mentioned in #18609, Windows downloads the certificates on-demand. Running `certutil.exe -syncWithWU` would download all of those certificates.

I tested it on my local system. Using the earlier implementation (which was reverted), I got around 30-40 certificates. Running `certutil.exe -syncWithWU` gave me around 370 certificates.

Reply:
Also, I'm not sure if the certificates which I get are exhaustive or not. They are downloaded from windows update. But I can't say for sure if any certificate might be missing.